crates/orchestrator/src/network/node.rs

```diff
@@ -355,6 +412,101 @@ impl NetworkNode {
+    ///
+    /// # Arguments
+    /// * `substring` - The substring or pattern to match within log lines.
+    /// * `is_glob` - Whether to treat `substring` as a glob pattern (`true`) or a regex (`false`).
+    /// * `options` - Configuration for timeout, match count predicate, and full-duration waiting.
+    ///
+    /// # Returns
+    /// * `Ok(LogLineCount::TargetReached(n, d))` if the predicate was satisfied within the timeout,
+    /// * `Ok(LogLineCount::TargetFails(n))` if the predicate was not satisfied in time,
+    /// * `Err(e)` if an error occurred during log retrieval or matching.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use std::sync::Arc;
+    /// # use provider::NativeProvider;
+    /// # use support::{fs::local::LocalFileSystem};
+    /// # use zombienet_orchestrator::{Orchestrator, network::node::{NetworkNode, LogLineCountOptions}};
+    /// # use configuration::NetworkConfig;
+    /// # async fn example() -> Result<(), anyhow::Error> {
+    /// #   let provider = NativeProvider::new(LocalFileSystem {});
+    /// #   let orchestrator = Orchestrator::new(LocalFileSystem {}, provider);
+    /// #   let config = NetworkConfig::load_from_toml("config.toml")?;
+    /// #   let network = orchestrator.spawn(config).await?;
+    /// let node = network.get_node("alice")?;
+    /// // Wait (up to 10 seconds) until pattern occurs once
+    /// let options = LogLineCountOptions {
+    ///     predicate: Arc::new(|count| count == 1),
+    ///     timeout_secs: 10,
+    ///     wait_until_timeout_elapses: false,
+    /// };
+    /// let result = node
+    ///     .wait_log_line_count_with_timeout_v2("error", false, options)
+    ///     .await?;
+    /// #   Ok(())
+    /// # }
+    /// ```
+    pub async fn wait_log_line_count_with_timeout_v2(
+        &self,
+        substring: impl Into<String>,
+        is_glob: bool,
+        options: LogLineCountOptions,
+    ) -> Result<LogLineCount, anyhow::Error> {
+        let secs = options.timeout_secs;
+        let substring = substring.into();
+        debug!("waiting until match lines count within {secs} seconds");
+
+        let start = tokio::time::Instant::now();
+
+        let match_fn: Box<dyn Fn(&str) -> bool + Send + Sync> = if is_glob {
+            Box::new(move |line: &str| glob_match(&substring, line))
+        } else {
+            let re = Regex::new(&substring)?;
+            Box::new(move |line: &str| re.is_match(line))
+        };
+
+        let deadline = Duration::from_secs(secs);
+
+        if options.wait_until_timeout_elapses {
+            tokio::time::sleep(Duration::from_secs(secs)).await;
+        }
+
+        let mut q;
+        loop {
+            q = 0_u32;
+            let logs = self.logs().await?;
+            for line in logs.lines() {
+                if match_fn(line) {
+                    q += 1;
+                    if !options.wait_until_timeout_elapses && (options.predicate)(q) {
```

Comment:
if `options. wait_until_timeout_elapses` is true we already _waited_ the desired time so we can just use 

```suggestion
                    (options.predicate)(q) {
```

Reply:
That's true. But on the other hand if `options.wait_until_timeout_elapses` is true I thought we should check the condition after processing all available lines. Otherwise we could return `TargetReached` upon first time when predicate is true ignoring that in the next steps the predicate might become false.
Example:
- we search for exactly one line "system launched", we shall wait until timeout 2s elapses.
So we waited 2s before the loop and then in the loop we are processing the the log (eg. 1000 lines).
In line 100 we found the "system launched" and we immediately return `TargetReached` ignoring the rest of the log, which could include another "system launched" in line 900. So in total there are 2 lines matching the pattern but we expected exactly one.

I will add some comment to make it more clear.

Reply:
ahh make sense, good catch!